---
title: "RNA-seq analysis in R"
subtitle: "Differential Expression of RNA-seq data"
author: "Stephane Ballereau, Mark Dunning, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
  html_document:
    toc: yes
minutes: 300
layout: page
bibliography: ref.bib
---

# Recap of pre-processing

The previous section walked-through the pre-processing and transformation of the
count data. Here, for completeness, we list the minimal steps required to 
process the data prior to differential expression analysis.

```{r recap, eval = FALSE}

# Read the sample information into a data frame
sampleinfo <- read_tsv("data/SampleInfo.txt")

# Read the data into R
seqdata <- read_tsv("data/GSE60450_Lactation.featureCounts", comment = "#")

# Transform the data to matrix of counts
countdata <- as.data.frame(seqdata) %>% 
    column_to_rownames("Geneid") %>% # turn the geneid column into rownames
    rename_all(str_remove, ".bam") %>% # remove the ".bam" from the column names
    select(sampleinfo$Sample) %>% # keep sample columns using sampleinfo
    as.matrix()

# filter the data to remove genes with few counts
keep <- rowSums(countdata) > 5
countdata <- countdata[keep,]
```

## Load the data

Alternatively we can load the `objects with the Rdata file we created in the 
pre-processing tutorial.

```{r loadData}
# before proceeding clean all objects out of the environment
rm(list=ls())

# load the Rdata object we created in the previous session
load("Robjects/preprocessing.Rdata")
ls()
dim(countdata)
sampleinfo
```

# Differential expression with DESeq2

First load the packages we need.

```{r setup, message = FALSE}
library(tidyverse)
library(DESeq2)
```

Now that we are happy that that the quality of the data looks good, we can 
proceed to testing for differentially expressed genes. There are a number of 
packages to analyse RNA-Seq data. Most people use 
[DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 
or [edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html). They 
are both equally applicable. There is an informative and honest blog post
[here](https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/) by Mike Love,
one of the authors of DESeq2, about deciding which to use.

We will use **DESeq2** for the rest of this practical.

## Create a DESeqDataSet object with the raw data

### Creating the design model formula

First we need to create a design model formula for our analysis. DESeq2 will 
use this to generate the model matrix, as we have seen in the linear models 
lecture. 

We have two variables: "status"" and "cell type". We will fit two models under 
two assumptions: no interaction and interaction of these two factors. 

Let's start with the model with only main effects, that is no interaction. 
The main assumption here is that the effect of the status is the same in both 
type of cells.

```{r modelForumla}
# Use the standard R 'formula' syntax for an additive model
design <- as.formula(~ CellType + Status)
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix
# Let's set factor levels on Status to use virgin as the intercept
sampleinfo$Status <- factor(sampleinfo$Status, 
                              levels = c("virgin", "pregnant", "lactate"))
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix

```

### Build a DESeq2DataSet

We don't actually need to pass DESeq2 the model matrix, if we pass it the design
formula it will build the matrix itself.

```{r makeDDSObj}
# create the DESeqDataSet object
ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = design)
```

## Data exploration

Let's plot a PCA from `vst` transformed data. 
Can you anticipate if the interaction term will be important?

```{r pcaPlot}
vstcounts <- vst(ddsObj, blind=TRUE)
plotPCA(vstcounts, intgroup=c("Status", "CellType"))
```

## DESeq workflow

The main DESeq2 work flow is carried out in 3 steps:

First, Calculate the "median ratio" normalisation size factors...

```{r commonSizeFactors}
ddsObj <- estimateSizeFactors(ddsObj)
```

... then estimate dispersion ...

```{r genewiseDispersion}
ddsObj <- estimateDispersions(ddsObj)
```

... finally, apply Negative Binomial GLM fitting and calculate Wald statistics
```{r applyGLM}
ddsObj <- nbinomWaldTest(ddsObj)
```

## The `DESeq` command

In practice the 3 steps above can be performed in a single step using the 
`DESeq` wrapper function. Performing the three steps separately is useful if you
wish to alter the default parameters of one or more steps, otherwise the `DESeq`
function is fine.

```{r theShortVersion}
# rebuild a clean DDS object
ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = design)
# Run DESeq
ddsObj <- DESeq(ddsObj)
```

## Generate a results table

We can generate a table of differential expression results from the DDS object
using the `results` function of DESeq2. DESeq2 will extract the 

From DESeq2 manual:
"The results function of the DESeq2 package performs independent filtering by default using the mean of normalized counts as a filter statistic. A threshold on the filter statistic is found which optimizes the number of adjusted p values lower than a [specified] significance level".
The default significance level for independent filtering is `0.1`, however, you
should set this to the FDR cut off you are planning to use. We will use `0.05`

```{r resultsTable}
res <- results(ddsObj, lfcThreshold = 0.05)
head(res)
```

Let's have a look at the model matrix to understand why DESeq2 has given us this
particular contrast.

```{r veiwModelMatrix}
modelMatrix
```

DESeq2 has the command `resultsNames` that allows us to view the contrast that
are available directly from the model matrix.

```{r resultsNames}
resultsNames(ddsObj)
```

Let's just rename `res` so that we know which contrast results it contains.

```{r}
resLvV <- res
```


```{r resultPvV}
# get the results for pregant versus virgin
resPvV <- results(ddsObj, 
                  name="Status_pregnant_vs_virgin", 
                  lfcThreshold = 0.05)
resPvV
```

Let's get the top 100 genes by adjusted p-value

```{r topGenesPvV, message = F}
topGenesPvV <- resPvV %>%
    as.data.frame() %>% 
    filter(!is.na(padj)) %>% 
    arrange(padj) %>% 
    head(100)
head(topGenesPvV)
```

> ### Challenge 1 {.challenge}
> Obtain results for basal vs luminal and find the top 200 genes (call the new
> results object `resBvL`.

**Solution**
```{r solutionChallenge1}
# resBvL <- results(ddsObj, name="CellType_basal_vs_luminal")
# topGenesBvL <- resBvL %>%
#     as.data.frame() %>%
#     filter(!is.na(padj)) %>%
#     arrange(padj) %>%
#     head(200)
# head(topGenesBvL)
```

## Contrasts

Suppose we want to find differentially expressed genes between **pregnant** and 
**lactate**. We don't have a parameter that explicitly will allow us to test 
that hypothesis. We need to provide a contrast.

```{r makeContrast}
resultsNames(ddsObj)

resPvL <- results(ddsObj,
                  contrast=c("Status", "pregnant", "lactate"),
                  lfcThreshold = 0.05)
```

## Comparing two design models

```{r compareModels}
designC <- as.formula(~ CellType )

# Compare the designs
ddsObjC <- DESeq(ddsObj, test="LRT", reduced=designC)
resCvCS <- results(ddsObjC)
```

> ### Challenge 2 {.challenge}
>
> 1.Fit a model with interaction: What is the rationale to include the 
> interaction?  
> 2. Is the number of replicates good enough to include the interaction?  
> 3. Is the interaction needed in the model?  


**Solution**
```{r solutionChallenge2}
# designI <- as.formula(~ CellType * Status)
# 
# # Build model
# ddsObjI <- DESeqDataSetFromMatrix(countData = countdata,
#                               colData = sampleinfo,
#                               design = designI)
# # Run DESeq
# ddsObjI <- DESeq(ddsObjI)
# 
# # Compare the models
# ddsObjI <- DESeq(ddsObjI, test="LRT", reduced=design)
# resIvA <- results(ddsObjI)
```


## Finally save the results in a new Rdata object

```{r saveObjects, eval=FALSE}

save(resLvV, ddsObj, sampleinfo, file="results/DE.Rdata")

```

-----------
