---
title: "RNA-seq analysis in R"
author: "Stephane Ballereau, Mark Dunning, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
  html_document:
    toc: yes
minutes: 300
layout: page
subtitle: Gene Set Testing for RNA-seq
bibliography: ref.bib
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
library(tidyverse)
```

The list of differentially expressed genes is sometimes so long that its interpretation becomes cumbersome and time consuming. A common downstream procedure is gene set testing. It aims to understand which pathways or gene networks the differentially expressed genes are implicated in.

Various ways exist to test for enrichment of biological pathways.

# GSEA analysis

Gene Set Enrichment Analysis GSEA was tests whether a set of genes of interest, 
e.g. genes
[@Subramanian15545]. The software is distributed by the [Broad 
Institute](http://software.broadinstitute.org/gsea/index.jsp) and is freely 
available for use by academic and non-profit organisations. 

In addition to the GSEA software the Broad also provide a number of very well 
curated `gene set`s for testing against your data - the [Molecular Signatures 
Database (MSigDB)](http://software.broadinstitute.org/gsea/msigdb/index.jsp). 
Unfortunately, these are collections of human genes, however, these lists
have been translated to mouse equivalents by the Walter+Eliza Hall Institutes
Bioinformatics service and made avaialble for [download](http://bioinf.wehi.edu.au/software/MSigDB/).


The analysis is performed by:

- (i) ranking all genes in the `data set`
- (ii) identifying the rank positions of all members of the `gene set` in the 
ranked `data set`
- (iii) calculating an enrichment score (ES) that represents the difference 
between the observed rankings and that which would be expected assuming a random 
rank distribution.

[commentary on GSEA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1266131/). 
The article describing the original software is available 
[here](http://www.pnas.org/content/102/43/15545.long).

## `fgsea`

The `fgsea` package [@Sergushichev2016] implements the same algorithm in R [vignette](http://www.bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html) "fast preranked gene set enrichment analysis (GSEA)".

```{r}
library(fgsea)
```

```{r loadData}
load("Robjects/Annotated_Results_LvV.RData")
```

## Create ranks

Rank all genes based on their fold change. We need to exclude genes for
which we do not have Entrez IDs. Also, we should use the shrunk LFC values.

```{r}
gseaDat <- shrinkLvV %>% 
    filter(!is.na(Entrez)) %>% 
    arrange(desc(logFC))
ranks <- gseaDat$logFC
names(ranks) <- gseaDat$Entrez
head(ranks)
```

Plot the ranked fold changes.

```{r}
#plot(ranks)
barplot(ranks)
```

## Load pathways

```{r}
load("data/mouse_H_v5.rdata")
pathways <- Mm.H
```

## Conduct analysis

```{r}
fgseaRes <- fgsea(pathways, ranks, minSize=15, maxSize = 500, nperm=1000)
#head(fgseaRes)
```

## Table of resLvV
```{r}
head(fgseaRes[order(padj, -abs(NES)), ], n=10)
```

## Enrichment score plot

```{r}
plotEnrichment(pathways[["HALLMARK_ESTROGEN_RESPONSE_EARLY"]], ranks)
```

Remember to check the [GSEA article](http://www.pnas.org/content/102/43/15545.full) for the complete explanation.

## GSEA table plot

The function `plotGseaTable` allows us to plot a summary figue showing the 
results for multiple pathways.

```{r}
topUp <- fgseaRes %>% 
    filter(ES > 0) %>% 
    top_n(10, wt=-padj)
topDown <- fgseaRes %>% 
    filter(ES < 0) %>% 
    top_n(10, wt=-padj)
topPathways <- bind_rows(topUp, topDown) %>% 
    arrange(-ES)

plotGseaTable(pathways[topPathways$pathway], 
              ranks, 
              fgseaRes, 
              gseaParam = 0.5)
```

# GO enrichment analysis

## `goseq`

GOseq is a method to conduct Gene Ontology (GO) analysis suitable for RNA-seq data as it accounts for the gene length bias in detection of over-representation [[@Young2010]](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14)

From the [GOseq vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf):

- GOseq first needs to quantify the length bias present in the dataset under consideration.
- This is done by calculating a Probability Weighting Function or PWF which can be thought of as a function which gives the probability that a gene will be differentially expressed (DE), based on its length alone.
- The PWF is calculated by fitting a monotonic spline to the binary data series of differential expression (1=DE, 0=Not DE) as a function of gene length.
- The PWF is used to weight the chance of selecting each gene when forming a null distribution for GO category membership.
- The fact that the PWF is calculated directly from the dataset under consideration makes this approach robust, only correcting for the length bias present in the data.

"GO analysis of RNA-seq data requires the use of random sampling in order to generate a suitable null distribution for GO category membership and calculate each category's significance for over representation amongst DE genes. ... In  most  cases,  the  Wallenius distribution can be used to approximate the true null distribution, without any significant loss in accuracy. The goseq package implements this approximation as its default option."

```{r goSeqPackage, message=FALSE}
library(goseq)
supportedOrganisms() %>% filter(str_detect(Genome, "mm"))
```

## Create a list of differentially expressed genes

```{r getDEGs}
genes <- as.integer(shrinkLvV$FDR < 0.01 & !is.na(shrinkLvV$FDR))
names(genes) <- shrinkLvV$GeneID
```

## Fit the Probability Weighting Function (PWF)

```{r pwFunction}
pwf <- nullp(genes, "mm10", "ensGene", bias.data = shrinkLvV$medianTxLength)
```

## Conduct GO enrichment analysis

```{r runGoseq}
goResults <- goseq(pwf, "mm10","ensGene", test.cats=c("GO:BP"))
```

## Plot the top 10

```{r plotGO}
goResults %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(`hits (%)`=numDEInCat*100/numInCat) %>% 
    dplyr::rename(Count=numDEInCat, p.value=over_represented_pvalue) %>% 
    ggplot(aes(x=`hits (%)`, y=term, colour=p.value, size=Count)) +
        geom_point() +
        geom_vline(xintercept = 0, linetype=2, alpha=0.3)
```

## Get the GO information for the GO accessions

```{r getGOinfo}
library(GO.db)
GOTERM[[goResults$category[1]]]
```

> ## Challenge 1 {.challenge}
>
> 1. Create a gene list for genes that are up-regulated by at least 4x (logFC>2)
> in lactating mice
> 2. Run a `goseq` analysis on this gene list
> 3. Plot the results
> 4. How is this result different to the previous GO analysis?

Look at genes that are up regulated by at least 4x (logFC>2) in lactating mices

```{r solution1, eval=F}
# isSigGeneUp <- shrinkLvV$FDR < 0.01 & 
#     !is.na(shrinkLvV$FDR) & 
#     shrinkLvV$logFC > 2
# 
# genesUp <- as.integer(isSigGeneUp)
# names(genesUp) <- shrinkLvV$GeneID
# 
# pwf <- nullp(genesUp, "mm10", "ensGene", bias.data = shrinkLvV$medianTxLength)
# 
# goResultsUp <- goseq(pwf, "mm10","ensGene", test.cats=c("GO:BP"))
# 
# goResultsUp %>% 
#     top_n(10, wt=-(over_represented_pvalue)) %>% 
#     mutate(`hits (%)`=numDEInCat*100/numInCat) %>% 
#     dplyr::rename(Count=numDEInCat,  p.value=over_represented_pvalue) %>% 
#     ggplot(aes(x=`hits (%)`, 
#                y=term, 
#                colour=p.value, 
#                size=Count)) +
#         geom_point() +
#         geom_vline(xintercept = 0, linetype=2, alpha=0.3)
```

# KEGG pathway enrichment analysis

## `clusterProfiler`

We can analyse for enrichment of KEGG pathways in much the same way as for GO 
terms. We could also use `goseq` for this, but this time we're going to use 
`clusterProfiler` [@Yu2012]. `clusterprofiler` supports direct online access of
the current KEGG database, rather than relying on R annotation packages, it
also provides some nice visualisation options.

```{r loadClusterProfiler}
library(clusterProfiler)
search_kegg_organism('mmu', by='kegg_code')
```

## KEGG enrichment analysis

```{r enrichKEGG}
sigGenes <- na.exclude(shrinkLvV$Entrez[ shrinkLvV$FDR < 0.01 & 
                                             !is.na(shrinkLvV$FDR) &
                                             abs(shrinkLvV$logFC) > 1 ])
kk <- enrichKEGG(gene         = sigGenes,
                 organism     = 'mmu')
head(kk, n=10)
```

## Visualise a pathway

### In a browser

`clusterProfile` has a function `browseKegg` that allows you to view the Kegg
pathway with the genes colours in in your browser.

```{r browseKegg}
browseKEGG(kk, 'mmu03320')
```

### As a file

The package `pathview` [@Luo2013] can be used to generate figures of KEGG 
pathways. One advantage over `clusterProfiles` browser methos is that the genes
are coloured according to their expression levels in our data. The package plots
the KEGG pathway to a `png` file in the working directory.

```{r pathview}
library(pathview)

logFC <- annotLvV$logFC
names(logFC) <- annotLvV$Entrez
pathview(gene.data  = logFC,
         pathway.id = "mmu03320",
         species    = "mmu",
         limit      = list(gene=5, cpd=1))
```

